Erstellen von Splitbuchungen

lenilsas opened the pull request:
neue Funktion um Buchungen zu Spliten

Commit:
Erstellen von Splitbuchungen

willuhn commented:
Vielen Dank für das Patch. Werde aber einige Tage brauchen, um mir das in Ruhe anzuschauen, da es ein komplexes Thema ist. Insbesondere zwei Bereiche sind mir da wichtig:

- Wenn eine Buchung gesplittet wird, existiert danach sowohl die ursprüngliche Buchung als auch die darauf verweisenden Splits. Ist hier sichergestellt, dass alle Berechnungen (Salden, Summen, Auswertungen, etc.) noch korrekt sind oder erscheinen dann dort zu hohe Summen, da die Beträge in Summe doppelt einfliessen?
- Ist die Performance beim Laden von Buchungslisten noch gewährleistet. Gerade in den letzten Releases habe ich daran gearbeitet, das Laden der Buchungsliste soweit zu optimieren, dass im Wesentlichen nur 3-4 SQL-Statements ausgeführt werden und die Anzahl der SQL-Queries nicht abhängig von der Anzahl der Buchungen ist. Gerade wenn die Datenbank nicht im LAN sondern remote per VPN angebunden wurde, ist das wichtig. Hier muss sichergestellt sein, dass beim Laden von Listen nicht irgendwo für jeden einzelnen Datensatz nochmal ein extra SQL-Query ausgelöst wird. Das kann man gut sehen, wenn man das Loglevel unter Datei->Einstellungen auf DEBUG stellt. Wenn es sich per SQL nicht lösen lässt, muss man da ggf. mit Caches arbeiten.

Hattest du schon Gelegenheit, die beiden Themen zu betrachten?

lenilsas commented:
- Bei den Gesplitteten Buchungen werden immer die Hauptbuchungen rausgefiltert und nur die aufgespitteten Teile Summiert. Dass habe ich in allen Listen und Auswertungen geprüft.
- Im Moment werden noch mehrere SQL Querries pro Buchung durchgeführt. Das schaue ich mir nochmal an, ob ich das mit einem Querry oder mit Cache lösen kann.

Commit:
Verbesserungen bei Splitbuchungen damit nicht so viele SQL anfragen
gestellt werden

lenilsas commented:
Ich habe ein paar Anpassungen vorgenommen, damit es nicht so Viele SQL Queries gibt. So sollte es von der Performance her wesentlich besser sein.

willuhn commented on src/de/willuhn/jameica/fibu/server/KontoImpl.java:
```diff
@@ -226,14 +226,13 @@ private double getUmsatz(DBIterator buchungen, DBIterator hilfsbuchungen) throws
     int kontoArt = getKontoArt().getKontoArt();
     Kontotyp typ = getKontoTyp();
+    //Split-Hauptbuchungen rausfiltern
+    buchungen.addFilter("NOT EXISTS(SELECT 1 FROM buchung b WHERE b.split_id = buchung.id)");
```
Hier wird der DBIterator von außen als Parameter reingereicht. Daher sollten hier keine weiteren SQL-WHERE-Bedingungen hinzugefügrt werden. Das Risiko ist zu hoch, dass das irgend Nebeneffekte hat.

Da aber ohnehin nur server-seitig über die Buchungen iteriert wird, ohne einzelne Buchungen an die UI zu übertragen, kann man die unerwünschten Buchungen hier auch einfach überspringen.

Wenn ich das richtig verstanden habe, sollte ja der Betrag der Hauptbuchung identisch mit der Summe der Split-Buchungen sein, oder? In dem Fall könnte man hier ja umgekehrt die Splitbuchungen weglassen und nur die Hauptbuchungen addieren. Etwa so:

```java
    while (buchungen.hasNext())
    {
      Transfer t = (Transfer) buchungen.next();

      // "getSplitID()" müsste man ggf. noch hinzufügen.
      // Das sollte doch dann die Split-Buchungen ignorieren.
      if ((t instanceof Buchung) && ((Buchung)t).getSplitID() != null)
        continue;

      if (t.getSollKonto().equals(this))
        soll += t.getBetrag();
      else
        haben += t.getBetrag();
    }
```

willuhn replied:
Stimmt. Das mit der Einschränkung auf das Konto ist natürlich korrekt. Da macht das hier Sinn.

willuhn commented on src/de/willuhn/jameica/fibu/io/report/VelocityReportBuchungList.java:
```diff
@@ -55,14 +55,12 @@ protected VelocityReportData getData(ReportData data) throws Exception
     
     list.setOrder("order by datum");
+    //Split-Hauptbuchungen rausfiltern
+    list.addFilter("NOT EXISTS(SELECT 1 FROM buchung b WHERE b.split_id = buchung.id)");
```
Sowohl hier als auch in VelocityReportKontoAuszug kommt "list" aus "jahr.getHauptBuchungen(....)". Könnte man das zusätzliche Filterkriterium "NOT EXISTS(SELECT 1 FROM buchung b WHERE b.split_id = buchung.id)" nicht direkt dort mit unterbringen? Oder ggf. "jahr.getHauptBuchungen(...)" um einen Boolean-Parameter "noSplitHauptbuchungen" erweitern? Dann hätte man das an zentraler Stelle und kann sicher sein, dass bei allen Aufrufern korrekt berücksichtigt ist, ob die Split-Hauptbuchungen enthalten sein sollen oder nicht.

willuhn commented on src/de/willuhn/jameica/fibu/util/BuchungUtil.java:
```diff
@@ -116,4 +119,96 @@ public Object extract(ResultSet rs) throws RemoteException, SQLException
+	      {
+	        DBService service = (DBService) Application.getServiceFactory().lookup(Fibu.class,"database");
+
+	        String sql = "SELECT id,(SELECT count(id) FROM buchung WHERE split_id = b.id) FROM buchung b WHERE geschaeftsjahr_id = ?";
```
Wenn ich das richtig verstehe, geht es doch nur darum eine Map zurückzuliefern, die alle Splitbuchungen enthält. Alle Buchungen, die nicht in der Liste enthalten sind, sind damit ja implizit entweder die Hauptbuchungen oder ungesplittete Buchungen. In dem Fall würde ich den Ergebnis-Typ auf ein HashSet umstellen, welches nur noch die IDs der Splitbuchungen enthält. Das spart dann auch das Sub-Select.

Also etwa so: "SELECT id FROM buchung WHERE split_id IS NOT NULL".

Und dann einfach ein Set mit diesen IDs zurückliefern.

willuhn reviewed (changes requested):
Ich habe ein paar Anmerkungen hinterlassen. Kannst du dir das mal bitte anschauen?

Commit:
Filtern der SplitHauptBuchungen in die Konto und Geshaeftsjahr Klassen
verschoben, hier zusaetzlicher parameter "noSplitHauptbuchungen" in
getHauptBuchungen

Commit:
BuhungUtil isSplit von Map zu Set geändert, SQL vereinfacht

willuhn commented:
Danke für die Änderungen. Ich schau's mir. Kann aber noch ein paar Tage dauern.

willuhn commented on src/de/willuhn/jameica/fibu/gui/action/BuchungSplitNeu.java:
```diff
@@ -0,0 +1,87 @@
+	        	Buchung b = (Buchung) i.next();
+		       betrag +=  b.getBruttoBetrag();
+	        }
+	        double betrag_alt = Double.parseDouble("0"+hauptBuchung.getKommentar());
```
Das hier verstehe ich nicht. Warum der Kommentar? Selbst wenn da nichts drin steht, erzeugt das ein "0null" und damit eine NumberFormatException.

lenilsas replied:
Ich habe das Kommentar Feld missbraucht um den ursprünglichen Betrag der Buchung zu speichern. Korrekter wäre natürlich das Betrag Feld zu verwenden, das ist jedoch komplizierter, da ich ja dann den Brutto-Betrag ohne Splitbuchungen holen muss. 
Um die Exception zu umgehen hatte ich in AbstractBaseBuchungImpl getKommentar() geändert, hatte ich beim Commit vergessen.
```
public String getKommentar() throws RemoteException
  {
	  Object s = this.getAttribute("kommentar");
	  if(s == null)
		  return "";
	  return (String) s;
  }
```

lenilsas replied:
Ich habe es nun so geändert, dass das Betrag Feld verwendet wird

lenilsas committed:
Bei Splitbuchungen Betrag feld statt Kommentar verwenden, Netto und
Brutto Betraege in Buchungsliste richtig anzeigen

lenilsas committed:
Das Filtern der SplitHauptbuchungen angepasst, so dass Funktionen mit
und ohne zusaetzlichem Parameter bestehen. Standartmaessig keine
SplitHauptbuchungen zurueckgeben

willuhn commented:
Danke für das Patch. Habe es übernommen.

mokraemer commented:
@lenilsas vielen Dank für das Feature - sowas habe ich mir schon lange gewünscht.
Leider gibt es aktuell noch keine Doku wie das ganze genau zu verwenden ist.
Man braucht aktuell ja eine primäre Buchung, die man danach wieder splitten kann - unklar ist von welchem auf welches Konto man dann die primäre Buchung macht. Oder ob beim Spiltten dann einfach die primäre Buchung bereits den Rest erhält.
Frage ist auch, ob man das aus der Buchung mit Hibiscus auch direkt erreichen könnte. Ich buche (korrekterweise) immer gegen das Bankkonto und muß dann halt immer bspw. Hotelrechnungen wg der MwSt oder Eigenanteil splitten. Wenn man das direkt machen könnte, wäre das eine echte Hilfe.

avl61 commented:
Hallo lenilsas, auch ich sage vielen Dank für die Funktion in Syntax. Nach Rücksprache mit Olaf , möchte auch ich die Verbesserung vorschlagen, die Funktion in das Kontextmenü aus Hibiscus (rechte Maustaste) mit einem Menüpunkt "als Splitbuchung in Syntax übernehmen" einzubauen. 
Liebe Grüße, A. Vogt

lenilsas commented:
Danke für die Rückmeldungen.

> @lenilsas vielen Dank für das Feature - sowas habe ich mir schon lange gewünscht. Leider gibt es aktuell noch keine Doku wie das ganze genau zu verwenden ist. Man braucht aktuell ja eine primäre Buchung, die man danach wieder splitten kann - unklar ist von welchem auf welches Konto man dann die primäre Buchung macht. Oder ob beim Spiltten dann einfach die primäre Buchung bereits den Rest erhält. Frage ist auch, ob man das aus der Buchung mit Hibiscus auch direkt erreichen könnte. Ich buche (korrekterweise) immer gegen das Bankkonto und muß dann halt immer bspw. Hotelrechnungen wg der MwSt oder Eigenanteil splitten. Wenn man das direkt machen könnte, wäre das eine echte Hilfe.

Ja, die Doku von SynTAX fehlt ja leider bisher komplett.

Wenn man eine Splitbuchung anlegen möchte, muss man auf eine bestehende Buchung->Rechtsklick->"Buchung Aufteilen (Split)..." Dann kommt die "normale" Buchungseingabe mit alle Daten der Hauptbuchung. Unten steht der Hinweis "Dies ist eine Splitbuchung".  Bei klicken auf "Speichern und nächste Buchung" wir gespeichert und eine neue Buchung geöffnet mit den Daten der Hauptbuchung und dem fehlenden Betrag.
Wenn alle Splitbuchungen angelegt sind (Restbetrag 0) kann man zur Buchungsübesicht zurück gehen. Hier wird nun die gesplittete Buchung fett dargestellt, die Spalten mit den Konten sind leer, es wird nur die Hauptbuchung, nicht die Splitbuchungen angezeigt.
Beim öffnen der Hauptbuchung wird eine Liste mit allen Splitbuchungen angezeigt. Dort wird auch der "ursprüngliche Gesamtbetrag" angezeigt sowie die Summe der Splitbuchungen. Wenn manches Soll und manche Habebuchungen sind können diese Summen auch verschieden sein.
In der Buchungsliste bei den Konten werden nur die Splitbuchungen, nicht jedoch die Hauptbuchungen aufgelistet. Die Splitbuchungen werden kursiv dargestellt.
Auch in den Auswertungen werden nur die Splitbuchungen ausgewertet und aufgelistet. Daher sind die Konten der Hauptbuchung zweitrangig, sie tauchen nirgends mehr auf.

Mein Vorgehen ist so, dass ich die Buchungen in Hibiscus einer Kategorie zuordne und dann alle nach vordefinierten Zuordnungen nach Syntax übernehme. anschließend prüfe ich die Buchungen un splitte sie ggfs.
 Das direkte Splitten aus Hibiscus heraus war daher für mich nie nötig.
Ich kann jedoch nochmal schauen, ob ich das ermöglichen könnte. Ich habe jedoch noch ein paar andere Projekte daher wird es noch etwas dauern...

mokraemer commented:
@lenilsas Danke für die Erläuterung - dann verhält es sich so, wie ich es dachte.
Ich finde das Feature sehr hilfreich - wie gesagt, der Wechsel aus Hibiscus immer nach Syntax ist halt dann etwas blöd - klar ich kann mir dann mit einem Kto helfen wo ich alle Splitbuchungen ablege und danach die dann Splitte (das war bisher auch mein Vorgehen: 1800->1373 und dann 1373->6680,6674,2071 - also Split von Übernachtung und Verpflegung, Eigenanteil). Hier konnte ich das dann gleich in einem Arbeitsschritt machen, sauberer wäre es halt mit deinem Ansatz, durchbricht aber den Arbeitsfluss.
